### PR TITLE
docs(ml): newsfeed text-tagger SLM v1 build spec (plan only, no implementation)

### PR DESCRIPTION
**Pull request by joemccann**
- [x] Red/green TDD (failing test, then fix): n/a, docs-only. `scripts/tests/test_docs_contract.py` 60 passed locally; owners rule added for the implement paths.
- [x] Staged only the files this change owns (no `git add -A`)
- [x] Owner doc updated, or the commit message has `docs-skip: <reason>`: the spec is the owner; `docs/README.md` row and `docs/owners.json` rule `slm-tagger` added
- [x] Previous `main` deploy finished before this push: docs-only branch, no deploy triggered

## What this PR is

Plan step only (Fable plans, Grok writes). `docs/ml/newsfeed-slm-tagger.md` is an implementer-grade spec for a local QLoRA-tuned specialist that becomes ladder rung `slm-tagger` **before** the paid and subscription rungs, for the Market Ear **text** tagger contract only. No trainer, server, migration or ladder code is in this PR. Nothing is enabled in production. Do not merge without Joe.

Honesty label carried throughout: **Radon SLM tagger: internal specialist, not a SotA replacement for open research.** The dollar case is tens of dollars a year (`docs/research/radon-small-model.md`); v1 is justified as ladder pressure relief / Cerebras-last hedge, format determinism, latency only where hardware permits, and a reusable pattern for later heads (out of scope).

## Enhancement folded in (Joe via CoS, QLoRA guide)

Eight rules from [x.com/sairahul1/status/2100882424343265527](https://x.com/sairahul1/status/2100882424343265527) are adopted as hard requirements, tagged `[HR-1]` to `[HR-8]` in the spec and each has its own checkbox in the done-when list. The guide's example task and unverified "beats 70B" claims are ignored.

| HR | Rule | Where it lands |
|---|---|---|
| 1 | Three-arm bakeoff: A current ladder, B prompt-only `Qwen2.5-1.5B-Instruct`, C QLoRA 1.5B. Cutover only on `C WINS`: F1 vs A on `gold_human` with bootstrap CI lower bound > 0, cost **and** p95 latency vs A on the serve host, and beats B | F.1, F.3 (gates G0-G8), I.2 step 1, L item 5 |
| 2 | Default base `Qwen/Qwen2.5-1.5B-Instruct`; size up (3B) only on a quality-gate failure, down (0.5B) only on the memory gate | D.2 |
| 3 | Constrained output: grammar + `accept_slm_tags_payload` (exactly 3, normalised, all in the live taxonomy); `abstain:count` / `abstain:out_of_taxonomy` / `abstain:empty` fall through the ladder. Closed-vocabulary rung; only the ladder coins tags | H.3, H.4 |
| 4 | Time split asserted by test; publisher split if a second publisher ever enters | C.3 |
| 5 | Loss masked to the tags, no text objective, taxonomy read live at call time, policy stays in `tagger.js` rules | C.2, D.3, J |
| 6 | Gates are `invalid_rate`, micro-F1, macro-F1, `rare_label_recall`, latency, throughput; loss is a model-card fact only | F.2, F.3 |
| 7 | GGUF `llama-server` sidecar (vLLM noted as the GPU drop-in); new I.3 post-deploy monitor: invalid/empty, out-of-taxonomy drift, JS-divergence label shift vs train distribution, 1-in-20 ladder sampling, availability, latency; daily oneshot timer, exit 3 pages | G.1, I.1, I.3 |
| 8 | Training data is Turso `posts` only; manifest `sources == ["turso.posts"]` asserted | C.1, C.6, K |

## Trainer pin (Joe via CoS, docs-only amendment)

Default train harness is **[LLaMA-Factory](https://github.com/hiyouga/LlamaFactory)** (`llamafactory-cli train` / `export`) for QLoRA/LoRA on `Qwen/Qwen2.5-1.5B-Instruct`. Train harness only, not runtime: inference stays the GGUF `llama-server` sidecar (vLLM/MLX as documented alternatives) and `RADON_SLM_TAGGER_MODE=off` until `C WINS`. mlx-lm is demoted to a documented Mac Mini alternative on the same `train.jsonl`; Unsloth and Axolotl are not required. Honest hardware note in D.3: bitsandbytes 4-bit is CUDA-only, so the same YAML runs as bf16 LoRA on the Mini (MPS) and as true QLoRA only on a rented GPU; either way the adapter is merged onto the bf16 base for GGUF and gate G0 measures adapter-vs-GGUF parity. HR-1 to HR-8, eval, serve and ladder sections are unchanged.

## Chosen path (one of each)

| Layer | Choice | Why |
|---|---|---|
| Base model | `Qwen/Qwen2.5-1.5B-Instruct` (Apache 2.0), shared by arms B and C | native LLaMA-Factory template and mature llama.cpp paths; Q4_K_M about 1.1 GB. Llama 3.2 rejected on licence friction, Phi/Gemma on size, Qwen3.5 deferred to v2 |
| Train stack | LLaMA-Factory QLoRA/LoRA (rank 16, alpha 32, 3 epochs, lr 1e-4 cosine, `train_on_prompt: false`); bf16 LoRA on the Mac Mini, QLoRA form on a rented GPU | one YAML for both hosts, no model-API key in the loop; mlx-lm documented alternative |
| Export | `llamafactory-cli export` (merge onto bf16), `convert_hf_to_gguf.py`, `llama-quantize Q4_K_M`; G0 parity | never merge from a 4-bit base |
| Serve stack | `llama-server` CPU build, `127.0.0.1:8331`, json_schema grammar, greedy, `max_tokens 64` | one binary; Ollama rejected; vLLM needs a GPU |
| Host | train on the Mac Mini, serve on the Hetzner app host beside `radon-newsfeed.service` | localhost call, no cross-host dependency |
| Weights | private B2 prefix, `manifest.json` + `MODEL_CARD.md` in git, no binaries in git | no git-lfs, `deploy.sh` drift guard |

## Hardware target and the memory gate

Hetzner host RAM/vCPU is **not recorded anywhere current** in the repo. The unit is enabled only if the 14-day `host_metrics.mem_avail_mb` p05 exceeds resident size + 512 MB (about 2,048 MB for 1.5B Q4_K_M). Unit sketch: `MemoryMax=2400M`, `CPUQuota=150%`, `Nice=10`, `--parallel 1`, pinned in `installed-units.sha256`, covered by the existing unit watchdog. If G7 fails on latency alone, the answer is a hardware decision, not a retrain.

## Ladder integration (section H)

`slm-tagger` is never added to `MODEL_LADDER_ORDER` or `wired_providers()`. `complete_text_json` gains `providers=` (forwarded to `_run_ladder`, which already accepts it). Only `model_ladder_cli.py`, only when `accept == "tags"`, builds the order per `RADON_SLM_TAGGER_MODE`: `off` (default, byte-identical to today) / `shadow` / `prefer` / `primary` (SLM then nvidia then cerebras; subscription band kept for distill and reviewer). Pin test required that distill callers never see the rung even with URL and `primary` set.

## Eval baselines the implement PR must measure

B0 label-noise ceiling (ladder re-tag vs stored `tags_text`, 200 posts); the A/B/C table on `test.jsonl` (about 1,200) and `gold_human.jsonl` (200) with every F.2 metric, latency and throughput measured on the serve host, and the B3 cost row. Verdict line is exactly `C WINS` or `C FAILS <gates>`.

## Data gaps found on the cloud VM

- No `posts.json`, no `data/tag_taxonomy.json` (untracked since `f88b413a` 2026-07-15; 212 tags then), no Turso credentials. Corpus size must be counted on a credentialed host; last verified 3,474 posts (2026-07-20).
- Test fixtures hold 1 to 3 synthetic posts each; not usable as a corpus.
- `scripts/newsfeed/CLAUDE.md` and `AGENTS.md` still claim the taxonomy file is force-tracked; the implement PR corrects them.
- Training target is `posts.tags_text` (the text tagger's own column). `tags` is a union with vision tags plus parent tags and is not the contract; `research_post_sources` rows are excluded.

## Joe approval list (section L)

1. Mac Mini for about one hour of training plus the A/B/C eval runs (arm A is roughly 1,400 subscription calls, zero prepaid); confirm RAM.
2. Read `nproc`, `free -m`, 14-day `host_metrics` on the app host; if the gate fails: upsize, fall back to 0.5B (default proposal), or keep the rung off.
3. One hour labelling 200 posts via `review_cli.py` (the HR-1 win is judged on this slice).
4. Weights storage: prefix under the existing B2 archive bucket (default, no new secret) or a new bucket (new `RADON_MODELS_S3_*` secrets).
5. Sign off the bakeoff table (`C WINS` / `C FAILS`); no threshold is moved to manufacture a win.
6. Enable `shadow` in production.
7. Move to `prefer`, later `primary`, each after a 14-day window with the I.3 monitor green.
8. Merge the implement PR.
9. llama.cpp release pin in the approved tool pins.
10. Act on a monitor breach or retrain trigger after cutover.

Decided by the enhancement and no longer open: closed vocabulary for the SLM rung, default base, Turso-only training data. Default proposals need no new API keys, no prepaid spend, no cloud GPU.

<div><a href="https://cursor.com/agents/bc-2c88288e-7585-5c19-b1d4-cff9b51aa8a0?cursor_ref=pr_footer&cursor_cta=open_in_web"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-web-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-web-light.png">Open in Web</picture></a>&nbsp;<a href="https://cursor.com/background-agent?bcId=bc-2c88288e-7585-5c19-b1d4-cff9b51aa8a0&cursor_ref=pr_footer&cursor_cta=open_in_cursor"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-cursor-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-cursor-light.png">Open in Cursor</picture></a>&nbsp;</div>

### CODE REVIEW

**Comment by joemccann** [comment]
Folded into #532 (combined SLM tracking PR). The spec lives there at its #535 revision, pointed at the single corpus writer `scripts/slm/export_dataset.py` and the live-prompt rung policy. Closing; nothing lost.
